### PR TITLE
lombok: init at 1.16.8

### PR DESCRIPTION
**Pull request by CrystalGamma**
Tested on NixOS using chroot build.
This is my first package, is this OK or do I need to compile from sources (even though I think this library does a few bytecode hacks, so it would probably be difficult)?
Do I need to mention a maintainer in the metadata and if so, how do I find one?

### CODE REVIEW

File: pkgs/development/libraries/java/lombok/default.nix [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{stdenv, fetchurl}:
+
+stdenv.mkDerivation {
+  name = "lombok-1.16.8";
+  src = fetchurl {url = "https://projectlombok.org/downloads/lombok-1.16.8.jar"; sha256 = "0s7ak6gx1h04da2rdhvc0fk896cwqm2m7g3chqcjpsrkgfdv4cpy";};
```
Comment (joachifm) [inline pkgs/development/libraries/java/lombok/default.nix]: Small tip: re-using `name` in the url makes it easier to keep the two in sync.

**Comment by joachifm** [comment]
I suggest that you add yourself as the maintainer for this package (see `lib/maintainers.nix`).

Building from source is preferred, only use pre-built binaries if building from source is infeasible.

**Comment by CrystalGamma** [comment]
I've tried building the package from source, but it looks like the derivation is impure (I get different hashes each time I build).
Is this a normal thing with java libraries and can I do something about this?

File: pkgs/development/libraries/java/lombok/default.nix [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{stdenv, fetchurl}:
+
+stdenv.mkDerivation rec {
+  name = "lombok-1.16.8";
+  src = fetchurl {url = "https://projectlombok.org/downloads/${name}.jar"; sha256 = "0s7ak6gx1h04da2rdhvc0fk896cwqm2m7g3chqcjpsrkgfdv4cpy";};
```
Comment (joachifm) [inline pkgs/development/libraries/java/lombok/default.nix]: Typically, the desired format is
```
fetchurl {
  url = ...;
  sha256 = ...;
};
```
Doesn't matter that much, but maintaining consistency is nice.

File: pkgs/development/libraries/java/lombok/default.nix [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  phases = ["installPhase"];
+  installPhase = "mkdir -p $out/share/java; cp $src $out/share/java/lombok.jar";
+  meta = {
+    description = "A library that can write a lot of boilerplate for your Java project";
+    platforms = stdenv.lib.platforms.all;
+    license = stdenv.lib.licenses.mit;
+    homepage = https://projectlombok.org/;
+    maintainer = stdenv.lib.maintainers.CrystalGamma;
```
Comment (joachifm) [inline pkgs/development/libraries/java/lombok/default.nix]: This should be `maintainers = [ ... ]`.

File: pkgs/development/libraries/java/lombok/default.nix [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{stdenv, fetchurl}:
+
+stdenv.mkDerivation rec {
+  name = "lombok-1.16.8";
+  src = fetchurl {url = "https://projectlombok.org/downloads/${name}.jar"; sha256 = "0s7ak6gx1h04da2rdhvc0fk896cwqm2m7g3chqcjpsrkgfdv4cpy";};
+  phases = ["installPhase"];
+  installPhase = "mkdir -p $out/share/java; cp $src $out/share/java/lombok.jar";
+  meta = {
```
Comment (joachifm) [inline pkgs/development/libraries/java/lombok/default.nix]: A common idiom is to do `with stdenv.lib`; not important, but looks neater.

File: pkgs/development/libraries/java/lombok/default.nix [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{stdenv, fetchurl}:
+
+stdenv.mkDerivation rec {
+  name = "lombok-1.16.8";
+  src = fetchurl {url = "https://projectlombok.org/downloads/${name}.jar"; sha256 = "0s7ak6gx1h04da2rdhvc0fk896cwqm2m7g3chqcjpsrkgfdv4cpy";};
+  phases = ["installPhase"];
```
Comment (joachifm) [inline pkgs/development/libraries/java/lombok/default.nix]: Again, to maintain consistency, let list elements breathe.

**Comment by joachifm** [comment]
Could you paste what you tried? If building from source is more hassle than it's worth, we can merge what you have as it is now and worry about making it perfect later (it'd be nice if you could fix the superficial nitpicks, though).

**Comment by CrystalGamma** [comment]
What I tried was a simple Ant build:

```nix
{stdenv, fetchFromGitHub, jdk, ant}:

stdenv.mkDerivation  rec {
  name = "lombok-1.16.8";
  src = fetchFromGitHub {
    owner = "rzwitserloot";
    repo = "lombok";
    rev = "26ad9e7f5f4de77436d3d63e4a43e8cb77621c15";
    sha256 = "1jjjr5lf7fjaywf9grkb009al2prbg2dkcdj4bhjpnpx3b8gnc8k";
  };
  buildInputs = [ jdk ant ];
  buildPhase = "ant";
  installPhase = "mkdir -p $out/share/java; install -m444 dist/lombok.jar $out/share/java";
  meta = {
    description = "A library that can write a lot of boilerplate for your Java project";
    platforms = stdenv.lib.platforms.all;
    license = stdenv.lib.licenses.mit;
    homepage = https://projectlombok.org/;
    maintainers = [ stdenv.lib.maintainer.CrystalGamma ];
  };
}
```

I ran `sha256sum` on the resulting jar (is there some automated way to test for purity?), and it was different everytime. When I tried to use the `antBuild` function used by `junit`, I got build errors (makes sense, since `junit` is also `meta.broken = true`).

However, I just now found that `antBuild` uses a shellscript function called `canonicalizeJar` (from `build-support/release/functions.sh` which may do exactly what I need it to do. I'll try to integrate it now.

In other matters, the ant build invokes its own dependency manager „Ivy” … is it worth or even wanted to try and replace that with nix dependencies?

**Comment by CrystalGamma** [comment]
I tried canonicalizing on branch CrystalGamma/nixpkgs@canonicalize-lombok, but it's still impure … any other ideas?

**Comment by joachifm** [comment]
Let's just merge it and improve it later. Thanks

**Pull request merged**